File: day04/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

type (
	Pos struct {
		X int
		Y int
	}

	Board struct {
		Width    int
		Height   int
		Unmarked map[int]Pos
		Marked   map[int]Pos
		Rows     []int
		Cols     []int
	}
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	var nums []int
	var boards []*Board
	board := &Board{
		Height:   0,
		Unmarked: map[int]Pos{},
		Marked:   map[int]Pos{},
	}
	first := true
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		if first {
			first = false
			for _, i := range strings.Split(scanner.Text(), ",") {
				num, err := strconv.Atoi(i)
				if err != nil {
					log.Fatal(err)
				}
				nums = append(nums, num)
			}
			continue
		}
		line := scanner.Text()
		if line == "" {
			if board.Height > 0 {
				board.Rows = make([]int, board.Height)
				board.Cols = make([]int, board.Width)
				boards = append(boards, board)
				board = &Board{
					Height:   0,
					Unmarked: map[int]Pos{},
					Marked:   map[int]Pos{},
				}
			}
			continue
		}
		h := board.Height
		board.Height++
		row := strings.Fields(line)
		board.Width = len(row)
		for n, i := range row {
			num, err := strconv.Atoi(i)
			if err != nil {
				log.Fatal(err)
			}
			if _, ok := board.Unmarked[num]; ok {
				log.Fatalln("Duplicate num")
			}
			board.Unmarked[num] = Pos{
				X: n,
				Y: h,
			}
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	if board.Height > 0 {
		board.Rows = make([]int, board.Height)
		board.Cols = make([]int, board.Width)
		boards = append(boards, board)
		board = nil
	}

	skipMap := map[int]struct{}{}
	first = true
	for _, i := range nums {
		for n, j := range boards {
			if _, ok := skipMap[n]; ok {
				continue
			}
			k := markBoard(j, i)
			if k > -1 {
				if first {
					first = false
					fmt.Println("Part 1:", k*i)
				}
				skipMap[n] = struct{}{}
				if len(skipMap) >= len(boards) {
					fmt.Println("Part 2:", k*i)
					return
				}
			}
		}
	}
}

func markBoard(board *Board, num int) int {
	p, ok := board.Unmarked[num]
	if !ok {
		return -1
	}
	board.Marked[num] = p
	delete(board.Unmarked, num)
	board.Rows[p.Y]++
	board.Cols[p.X]++
	if board.Rows[p.Y] >= board.Height {
		return scoreBoard(board)
	}
	if board.Cols[p.X] >= board.Width {
		return scoreBoard(board)
	}
	return -1
}

func scoreBoard(board *Board) int {
	sum := 0
	for v := range board.Unmarked {
		sum += v
	}
	return sum
}

File: day04/src/main.rs
use std::collections::hash_map::Entry;
use std::collections::HashMap;
use std::collections::HashSet;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

struct Pos(usize, usize);

struct Board {
    width: usize,
    height: usize,
    unmarked: HashMap<i32, Pos>,
    marked: HashMap<i32, Pos>,
    rows: Vec<usize>,
    cols: Vec<usize>,
}

impl Board {
    fn new() -> Self {
        Board {
            width: 0,
            height: 0,
            unmarked: HashMap::new(),
            marked: HashMap::new(),
            rows: Vec::new(),
            cols: Vec::new(),
        }
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);
    let lines = reader.lines();

    let mut nums = Vec::new();
    let mut boards = Vec::new();
    let mut board = Board::new();
    let mut first = true;
    for line in lines {
        if first {
            first = false;
            nums = line?
                .split(",")
                .flat_map(|num| num.parse::<i32>())
                .collect::<Vec<_>>();
            continue;
        }
        let line = line?;
        if line == "" {
            if board.height > 0 {
                board.rows = vec![0; board.height];
                board.cols = vec![0; board.width];
                boards.push(board);
                board = Board::new();
            }
            continue;
        }
        let h = board.height;
        board.height += 1;
        let mut w = 0;
        for (n, i) in line.split_ascii_whitespace().enumerate() {
            w += 1;
            if let Entry::Vacant(v) = board.unmarked.entry(i.parse::<i32>()?) {
                v.insert(Pos(n, h));
            } else {
                return Err("Duplicate num".into());
            }
        }
        board.width = w;
    }
    if board.height > 0 {
        board.rows = vec![0; board.height];
        board.cols = vec![0; board.width];
        boards.push(board);
    }

    let mut skip_map = HashSet::new();
    first = true;
    for i in nums {
        let l = boards.len();
        for (n, j) in boards.iter_mut().enumerate() {
            if skip_map.contains(&n) {
                continue;
            }
            let k = mark_board(j, i);
            if k > -1 {
                if first {
                    first = false;
                    println!("Part 1: {}", k * i);
                }
                skip_map.insert(n);
                if skip_map.len() >= l {
                    println!("Part 2: {}", k * i);
                    return Ok(());
                }
            }
        }
    }

    Ok(())
}

fn mark_board(board: &mut Board, num: i32) -> i32 {
    if let Entry::Occupied(e) = board.unmarked.entry(num) {
        let (k, v) = e.remove_entry();
        let Pos(x, y) = v;
        board.marked.insert(k, v);
        board.rows[y] += 1;
        board.cols[x] += 1;
        if board.rows[y] >= board.height {
            return score_board(board);
        }
        if board.cols[x] >= board.width {
            return score_board(board);
        }
    }
    -1
}

fn score_board(board: &Board) -> i32 {
    board.unmarked.keys().sum()
}
